.. image:: https://img.shields.io/badge/License-MPL%202.0-brightgreen.svg
    :alt: License: MPL2.0
    :target: https://opensource.org/licenses/MPL-2.0

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :alt: Code style: black
    :target: https://github.com/python/black

.. image:: https://www.codefactor.io/repository/github/milescsmith/pygmst/badge?s=088f5e010a0d94e51cce34fe1fdcab54adc16734
    :target: https://www.codefactor.io/repository/github/milescsmith/pygmst
    :alt: CodeFactor

.. image:: https://coveralls.io/repos/github/milescsmith/pygmst/badge.svg?branch=master
    :target: https://coveralls.io/github/milescsmith/pygmst?branch=master

.. image:: https://travis-ci.org/milescsmith/pygmst.svg?branch=master
    :target: https://travis-ci.org/milescsmith/pygmst
    
pygmst
======

A Python translation of GeneMarkS-T from the Borodovsky group at the
Georgia Institute of Technology.

References:
-----------
Tang S. and Borodovsky M.
"GeneMarkS-T: identification of protein coding regions in RNA transcripts"

Other References for GeneMark Suite
Lukashin A. and Borodovsky M.
"GeneMark.hmm: new solutions for gene finding."
Nucleic Acids Research, 1998, Vol. 26, No. 4, pp. 1107-1115

Besemer J., Lomsadze A. and Borodovsky M.
"GeneMarkS: a self-training method for prediction of gene starts in
microbial genomes. Implications for finding sequence motifs in
regulatory regions."

Nucleic Acids Research, 2001, Vol. 29, No. 12, 2607-2618
Besemer J. and Borodovsky M.
"Heuristic approach to deriving models for gene finding."
Nucleic Acids Research, 1999, Vol. 27, No. 19, pp. 3911-3920
